File: core/src/commonMain/kotlin/work/socialhub/kmastodon/internal/_InternalUtility.java
package work.socialhub.kmastodon.internal;

import com.google.gson.Gson;
import mastodon4j.MastodonException;
import mastodon4j.entity.Error;
import mastodon4j.entity.share.Link;
import mastodon4j.entity.share.RateLimit;
import mastodon4j.entity.share.Response;
import net.socialhub.http.HttpException;
import net.socialhub.http.HttpRequestBuilder;
import net.socialhub.http.HttpResponse;
import net.socialhub.http.HttpResponseCode;

import java.net.URLEncoder;

/**
 * @author hecateball
 */
class _InternalUtility {

    private final static Gson gson = new Gson();

    private _InternalUtility() {
    }

    static String getBearerToken(String accessToken) {
        return "Bearer " + accessToken;
    }

    static Gson getGsonInstance() {
        return gson;
    }

    static Response<Void> proceed(RequestInterface function) {
        try {
            HttpResponse response = function.proceed();
            switch (response.getStatusCode()) {
                case HttpResponseCode.OK:
                    Response<Void> result = new Response<>();
                    result.setRateLimit(RateLimit.of(response));
                    return result;
                default:
                    Error error = gson.fromJson(response.asString(), Error.class);
                    throw new MastodonException(error, response.getStatusCode());
            }
        } catch (HttpException e) {
            throw new MastodonException(e);
        }
    }

    static <T> Response<T> proceed(Class<T> clazz, RequestInterface function) {
        try {
            HttpResponse response = function.proceed();
            switch (response.getStatusCode()) {
                case HttpResponseCode.OK:
                    Response<T> result = new Response<>();
                    result.set(gson.fromJson(response.asString(), clazz));
                    result.setRateLimit(RateLimit.of(response));
                    result.setLink(Link.of(response));
                    return result;

                default:
                    Error error = gson.fromJson(response.asString(), Error.class);
                    throw new MastodonException(error, response.getStatusCode());
            }
        } catch (HttpException e) {
            throw new MastodonException(e);
        }
    }

    interface RequestInterface<T> {
        HttpResponse proceed() throws HttpException;
    }

    static void addParam(HttpRequestBuilder builder, String key, Object value) {
        if (value != null) {
            builder.param(key, value.toString());
        }
    }

    static String encode(String str) {
        try {
            return URLEncoder.encode(str, "utf-8");
        } catch (Exception e) {
            return str;
        }
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/internal/_InternalUtility.kt
package work.socialhub.kmastodon.internal

import com.google.gson.Gson
import mastodon4j.MastodonException
import mastodon4j.entity.Error
import mastodon4j.entity.share.Link
import mastodon4j.entity.share.RateLimit
import mastodon4j.entity.share.Response
import net.socialhub.http.HttpException
import net.socialhub.http.HttpRequestBuilder
import net.socialhub.http.HttpResponse
import net.socialhub.http.HttpResponseCode

/**
 * @author hecateball
 */
internal object _InternalUtility {
    private val gson: Gson = Gson()

    fun getBearerToken(accessToken: String): String {
        return "Bearer $accessToken"
    }

    val gsonInstance: Gson
        get() = gson

    fun proceed(function: RequestInterface<*>): Response<java.lang.Void> {
        try {
            val response: HttpResponse = function.proceed()
            when (response.getStatusCode()) {
                HttpResponseCode.OK -> {
                    val result: Response<java.lang.Void> = Response()
                    result.setRateLimit(RateLimit.of(response))
                    return result
                }

                else -> {
                    val error: java.lang.Error = gson.fromJson(response.asString(), java.lang.Error::class.java)
                    throw MastodonException(error, response.getStatusCode())
                }
            }
        } catch (e: HttpException) {
            throw MastodonException(e)
        }
    }

    fun <T> proceed(clazz: java.lang.Class<T>?, function: RequestInterface<*>): Response<T> {
        try {
            val response: HttpResponse = function.proceed()
            when (response.getStatusCode()) {
                HttpResponseCode.OK -> {
                    val result: Response<T> = Response()
                    result.set(gson.fromJson(response.asString(), clazz))
                    result.setRateLimit(RateLimit.of(response))
                    result.setLink(Link.of(response))
                    return result
                }

                else -> {
                    val error: java.lang.Error = gson.fromJson(response.asString(), java.lang.Error::class.java)
                    throw MastodonException(error, response.getStatusCode())
                }
            }
        } catch (e: HttpException) {
            throw MastodonException(e)
        }
    }

    fun addParam(builder: HttpRequestBuilder, key: String?, value: Any?) {
        if (value != null) {
            builder.param(key, value.toString())
        }
    }

    fun encode(str: String): String {
        return try {
            java.net.URLEncoder.encode(str, "utf-8")
        } catch (e: java.lang.Exception) {
            str
        }
    }

    internal interface RequestInterface<T> {
        @Throws(HttpException::class)
        fun proceed(): HttpResponse
    }
}
